Environment Page Help Text

rebeccaalpert opened the pull request:
[Fixes #117](https://github.com/openshift/console/issues/117)

Added help text to Environment page. Shows only on pages that aren't read-only.

![screen shot 2018-06-11 at 1 57 35 pm](https://user-images.githubusercontent.com/7014965/41248668-bcab51d8-6d7f-11e8-8c35-bd5156f54ca8.png)

@openshift/team-ux-review, could you please review the design?

@stacymcauliffe - Chris & co. said I should ask you to verify the wording. Chris changed it slightly.

coreos-ui commented:
Can one of the admins verify this patch?

coreos-ui commented:
Can one of the admins verify this patch?

spadgett commented on frontend/public/components/environment.jsx:
```diff
@@ -191,6 +191,11 @@ export class EnvironmentPage extends PromiseComponent {
 
     return <div className="co-m-pane__body">
+      { !readOnly &&
+          <div>
```
Do we need this div or can we just use the `p` tab?

jcaianirh replied:
I think this div is unncessary.

spadgett commented on frontend/public/components/environment.jsx:
```diff
@@ -191,6 +191,11 @@ export class EnvironmentPage extends PromiseComponent {
     return <div className="co-m-pane__body">
+      { !readOnly &&
+          <div>
+            <p className="co-m-pane__explanation">Define environment variables as key-value pairs to store configuration settings. Enter text or reference a value from a ConfigMap or Secret. Drag and drop the environment variables to change the order they are run in. Lower variables can reference higher ones, for example FULLDOMAIN = $(SUBDOMAIN).example.com</p>
```
Maybe "Later variables can reference earlier ones"? The terms lower and higher confused me.

benjaminapetersen replied:
And perhaps a `<code>` around the code `FULLDOMAIN = $(SUBDOMAIN).example.com`?

spadgett commented:
jenkins retest

cshinn commented:
Styling LGTM

stacymcauliffe commented:
Agree with @spadgett about wording feeling a bit awkward. Can we update text: 

Define environment variables as key-value pairs to store configuration settings. You can enter text or add values from a ConfigMap or secret. 

Drag and drop environment variables to change the order in which they are run. A variable can reference any other variables that come before it in the list, for example `FULLDOMAIN = $(SUBDOMAIN).example.com`.

rebeccaalpert commented:
I removed the junk div, added the code tags and new text, and updated the code tag color/message and header spacing per Jenny Haines' feedback on the openshift-ux Slack channel.

![screen shot 2018-06-11 at 4 10 08 pm](https://user-images.githubusercontent.com/7014965/41302961-ef75e384-6e39-11e8-8ddd-9909623eb78d.png)

spadgett commented on frontend/public/components/environment.jsx:
```diff
@@ -184,13 +184,16 @@ export class EnvironmentPage extends PromiseComponent {
 
     return <div className="co-m-pane__body">
+      { !readOnly &&
+          <p className="co-m-pane__explanation environment">Define environment variables as key-value pairs to store configuration settings. You can enter text or add values from a ConfigMap or secret. Drag and drop environment variables to change the order in which they are run. A variable can reference any other variables that come before it in the list, for example <code>FULLDOMAIN = $(SUBDOMA ...
```
I would capitalize Secret to match ConfigMap.

spadgett commented on frontend/public/style/_common.scss:
```diff
@@ -107,7 +107,7 @@
   font-size: 20px;
   font-weight: 600;
-  margin: 0 0 30px 0;
+  margin: 0 0 20px 0;
```
This is going to change the margin on all the pages, even those without explanations. I'm not sure we want that.

rebeccaalpert replied:
@jennyhaines Do you want the spacing changes to affect all the headers (even when there is no description)? This reduces the spacing between a header and its content.

jennyhaines replied:
Oh, I see. No, I don't think that the margin should be changed in instances when there is no description. Is there a way to add an additional spacer between the description and the header without applying directly to the header or the description?

rebeccaalpert replied:
@jennyhaines I can add extra spacing between the description and header only when the description is there. I'll change this to reduce spacing between the header and content when the description is visible.

rebeccaalpert replied:
@jennyhaines I talked to Joe. If we reduce the space between “Container hello-openshift” and the form fields conditionally when the description is there, it'll make the site inconsistent. Do you want it changed globally instead?

I added the space between the instructional text and “Container hello-openshift" so it's only there when the instructional text is there.

spadgett commented on frontend/public/style/_common.scss:
```diff
@@ -137,7 +137,12 @@
+
+.co-m-pane__explanation code {
+  background-color: $co-gray-lightest;
+  color: $co-gray;
```
@jennyhaines Do we want to change the `code` style in only one spot?

I wonder if we should just not use `<code>` tags if we don't want the styling.

jennyhaines replied:
@spadgett @rebeccaalpert I actually don't see the need to have the code styling be so prominent with the blue anywhere. It seems too prominent within a block of text that is generally used as description text. Since this text is not meant to be more important than (or detract from) its subsequent form field or data display, I'd suggest changing the styling to grey everywhere rather than blue. I was looking at [this example](http://www.patternfly.org/pattern-library/widgets/#cnt-remaining-chars) on PatternFly of the `<code>` tag being used multiple times in one block, and it just seems like too much.

spadgett commented:
jenkins retest

spadgett commented:
jenkins rebuild

stevekuznetsov commented:
/test backend
/test frontend

spadgett commented:
Looks like you've accidentally pulled in some unrelated changes with the rebase

spadgett commented on frontend/public/style/_overrides.scss:
```diff
@@ -40,6 +40,10 @@ small {
   margin: 0 !important;
 }
 
+.margin-bottom {
```
This shouldn't be a generic class. I'd recommend

```css
.co-m-pane__explanation--margin-bottom {
```

and move it next to the `.co-m-pane__explanation` declaration. This is more in line with BEM naming. See http://getbem.com/naming/

I'm still a little uncertain why we need to special case this specific page. Can you add a before and after screenshot with the margin changes?

rebeccaalpert replied:
Jenny didn't like the visual hierarchy. She had asked for extra margins under the explanation (helptext) -- this class. She had additionally asked for the space between the "Container" header and form fields to be reduced. I had implemented the latter globally the last round of edits and you'd questioned it/she never responded as to whether she wanted that globally, so I took that change out.

With no margin changes:
![screen shot 2018-06-11 at 3 52 59 pm](https://user-images.githubusercontent.com/7014965/41621913-0e641656-73dc-11e8-80a9-0112949a2db8.png)

With margin changes Jenny wanted:
![screen shot 2018-06-11 at 4 10 08 pm](https://user-images.githubusercontent.com/7014965/41621916-0f076b26-73dc-11e8-8426-b318d68d2684.png)

tlwu2013 replied:
Hi @spadgett , 

After checked the screenshot, I feel I'm also in favor of having larger bottom-margin (30px) to have a better visual hierarchy. But instead of creating a new class in _overrides.scss, I am wondering if we can simply increase the margin-bottom of **co-m-pane__explanation** to **30px**? 

Ingress Details page:
[margin-bottom: 20px]
![ori __co-m-pane__explanation___margin-bottom_20px](https://user-images.githubusercontent.com/5903705/41625439-b388fe26-73cd-11e8-9556-a3ecc868537c.png)

[margin-bottom: 30px]
![new __co-m-pane__explanation___margin-bottom_30px](https://user-images.githubusercontent.com/5903705/41625445-b745c508-73cd-11e8-8f98-8dc906c0dc11.png)

Create RoleBinding page:
[margin-bottom: 20px]
![ori_create_binding __co-m-pane__explanation___margin-bottom_20px](https://user-images.githubusercontent.com/5903705/41625449-bb107b6a-73cd-11e8-818c-c0000682080f.png)

[margin-bottom: 30px]
![new_create_binding __co-m-pane__explanation___margin-bottom_30px](https://user-images.githubusercontent.com/5903705/41625456-bdcdedc4-73cd-11e8-9c10-b1004c8daa7d.png)


Also, I am a little bit worried about the help text already mentioned **... You can enter text or add values from a ConfigMap or secret. ** before users can actually "Add Value from Config Map or Secret" on this UI.

Are we going to have **Add Value from Config Map or Secret** for this release?

spadgett replied:
> Are we going to have Add Value from Config Map or Secret for this release?

We're working on it this sprint, but I agree we should probably just take the text out until it's done.

tlwu2013 replied:
Got it, yeah, I don't want to introduce the extra burden here. As long as we'll have the actual functionality to add value from ConfigMap/Secret, I am fine with either way. Please just go with the less problematic one, and thanks!

rebeccaalpert replied:
>But instead of creating a new class in _overrides.scss, I am wondering if we can simply increase the margin-bottom of co-m-pane__explanation to 30px?

Yeah, that's what I had in originally; I'll change it back.

What about the headline padding? Jenny wanted "Container hello-openshift" to have margin-bottom: 20px rather than 30px. Is it okay if that's a global change as well?

The class is .co-section-title.

tlwu2013 replied:
Chatted on Slack with @robszumski and @serenamarie125 , we like the margin-bottom: 20px.
The * .co-section-title* is also used on other section title. I've tried with a few titles and it seems fine. 

Let's do that.

*bottom-margin: 30px*
<img width="713" alt="bottom-margin__30px" src="https://user-images.githubusercontent.com/5903705/41791800-b09f544a-760b-11e8-8f1f-2bf97b79b040.png">
<img width="649" alt="screen shot 2018-06-22 at 10 58 56 am" src="https://user-images.githubusercontent.com/5903705/41791850-d289fe34-760b-11e8-9143-c3f69024d981.png">


*bottom-margin: 20px*
<img width="685" alt="bottom-margin__20px" src="https://user-images.githubusercontent.com/5903705/41791801-b0ca1dba-760b-11e8-9cb7-c5e7788ad3eb.png">
<img width="673" alt="screen shot 2018-06-22 at 10 58 41 am" src="https://user-images.githubusercontent.com/5903705/41791841-cc653e60-760b-11e8-9e1a-820a4d4af315.png">

rebeccaalpert replied:
Thanks!

spadgett commented on frontend/public/style/_overrides.scss:
```diff
@@ -40,6 +40,10 @@ small {
 }
 
+.margin-bottom {
+  margin-bottom: 30px !important;
```
Why is `!important` needed?

jennyhaines commented:
So sorry that I hadn’t responded @rebeccaalpert @spadgett about whether I thought it would be a good idea to increase the spacing to 30px in all instances. I think that the screenshot examples that @tlwu2013 shared look good with the change. (Thank you for sharing those!) I’d be in favor of changing across the board.

serenamarie125 commented:
@rebeccaalpert looks like you are good to go with the 30px - both Tony & Jenny are in agreement with that.  thanks!

spadgett commented:
jenkins rebuild
